text/0003-performance-data.md

```diff
@@ -0,0 +1,133 @@
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Currently there is Transport performance data and MLAPI performance data that is not being propagated up externally for any consumer to render or consume. The use cases for this would be useful for technologies such as a **Network Stats Overlay** or a **Profiler** (**Network Messages** and **Network Operation** need this data in the vanilla Unity Profiler). Users need a viable and low overhead way to obtain this data (such as RPCs sent this frame/tick) from MLAPI and utilize this data in the way that they want for their own metrics.
+
+# Motivation
+[motivation]: #motivation
+
+This is needed to provide both a default and user specific implementations for future Network Profilers and Network Stats overlays. The MLAPI source code can remain Open Source, and yet, the data being streamed can be hooked up to external and low level systems.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to allow first and third party rendering of the data we will need to allow both MLAPI and its transports to propagate data up the chain. 
+If the profiler flags are enabled the following will need to happen: 
+Transports will need to output the number of bytes being sent in the current tick and the latency. The data will be written out to a ProfilerTickData chunk. This data will be generic and transport agnostic.
+MLAPI will need to output the number of networked objects and the amount of RPCs being sent. This data will be added to the ProfilerTickData chunk.
+
+External Tools and APIs will be able to then get this fully written ProfilerTickData at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a struct, accessing its data will be as simple as accessing fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in ```NetworkManager.LateUpdate()``` for the MLAPI and ```Transport.Send()``` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data).
+
+```cs
+struct ProfilerTickData {
+    int tickID;
+
+    // mlapi-level
+    int numberOfRPCs;
+    ...
+
+    // transport-level
+    TransportProfilerData transportProfilerData;
+    ...
+}
+
+
+struct TransportProfilerData {
```

Comment:
I don't understand why to retrieve this data a change to each Transport is necessary.
Bytes/send received is just the actual byte buffers send/received without all the header overhead from the transport layer right?
Same for numberOfMessagesOutgoing. That's "Number of send calls from MLAPI to the transport" and not number of networked packets sent out after the transport batched them.

So if you ask me we can move all this code to the place where Transport.Send is called and don't change transports at all.

Reply:
> Bytes/send received is just the actual byte buffers send/received without all the header overhead from the transport layer right?

Not necessarily, there could be more messages going back and forth (keep-alive pings, other QoS packets etc.) which could be sent and received without explicit transport-API-level `Send()` or `Poll()`.

> Same for numberOfMessagesOutgoing. That's "Number of send calls from MLAPI to the transport" and not number of networked packets sent out after the transport batched them.

Again, not necessarily. We might end up sending more or less number of packets on the transport-side.

> So if you ask me we can move all this code to the place where Transport.Send is called and don't change transports at all.

Also, this would mean that we will ALWAYS expect that struct being filled or left unfilled but returned back on every `Send()` — what if we never call `Send()` in a transport tick frame?

Beyond that, retrieving this profile stats data struct will be over on its own API which could even be wrapped into `#if UNITY_NETWORK_PROFILER`-like macro and left out entirely from release builds.

I think we're going in the right direction overall but we should discuss and decide what that intermediary struct should look like.

Reply:
I think that's t ne problem with this RFC/interface. It's not clear about what exactly we are measuring. Should the transport be doing keep alive etc.

Often the transports lacks an API to properly return amount of network packets sent and how many bytes sent.
We already have this problem with GetCurrentRTT of the transport interface. There are many ways to define RTT it's not entirely clear what exactly this means and different Transports right now have very different implementations.

If we add metrics to an interface like Transport they should be clearly defined which currently is not the case.

Reply:
I also agree that we should have a better API design and a better RFC document here. Let's discuss this in a meeting together!

Reply:
Updated to be more explicit as to why MLAPI is responsible for aggregating and controlling the buffer/performance data